bus/routes.go

```diff
@@ -2281,3 +2280,198 @@ func (b *Bus) multipartHandlerListPartsPOST(jc jape.Context) {
 	}
 	jc.Encode(resp)
 }
+
+func (b *Bus) rhpFormHandler(jc jape.Context) {
+	// apply pessimistic timeout
+	ctx, cancel := context.WithTimeout(jc.Request.Context(), 15*time.Minute)
+	defer cancel()
+
+	// decode the request
+	var rfr api.RHPFormRequest
+	if jc.Decode(&rfr) != nil {
+		return
+	}
+
+	// validate the request
+	if rfr.EndHeight == 0 {
+		http.Error(jc.ResponseWriter, "EndHeight can not be zero", http.StatusBadRequest)
+		return
+	} else if rfr.HostKey == (types.PublicKey{}) {
+		http.Error(jc.ResponseWriter, "HostKey must be provided", http.StatusBadRequest)
+		return
+	} else if rfr.HostCollateral.IsZero() {
+		http.Error(jc.ResponseWriter, "HostCollateral can not be zero", http.StatusBadRequest)
+		return
+	} else if rfr.HostIP == "" {
+		http.Error(jc.ResponseWriter, "HostIP must be provided", http.StatusBadRequest)
+		return
+	} else if rfr.RenterFunds.IsZero() {
+		http.Error(jc.ResponseWriter, "RenterFunds can not be zero", http.StatusBadRequest)
+		return
+	} else if rfr.RenterAddress == (types.Address{}) {
+		http.Error(jc.ResponseWriter, "RenterAddress must be provided", http.StatusBadRequest)
+		return
+	}
+
+	// fetch gouging parameters
+	gp, err := b.gougingParams(ctx)
+	if jc.Check("could not get gouging parameters", err) != nil {
+		return
+	}
+	gc := gouging.NewChecker(gp.GougingSettings, gp.ConsensusState, gp.TransactionFee, nil, nil)
+
+	// send V2 transaction if we're passed the V2 hardfork allow height
+	var contract rhpv2.ContractRevision
+	if b.isPassedV2AllowHeight() {
+		// form the contract
```

Comment:
I'd replace this with a panic. `rhp2.FormV2Contract` will never exist. It will be a method on the `rhp4` client.

Reply:
Yeah I know I just wanted to avoid the panic but I guess it's fine. Will remove.